module3/projects/rails_engine/requirements.md

```diff
@@ -93,17 +99,86 @@ module Api
 end
 ```
 
-## 3a. ReST Endpoints
+#### Error Responses
+
+If the user causes an error for which you are sending a 400-series error code, the JSON body of the response should follow a similar JSON Spec of a `data` element, with a `nil` ID, and empty attributes.
+
+As an **EXTENSION**, customize the error message to use this format instead:
+
+```json
+{
+  "message": "your query could not be completed",
+  "errors": [
+    "string of error message one",
+    "string of error message two",
+    "etc"
+  ]
+}
+```
+
+You can customize the value of the `message` element, but the `message` element must be present.
+
+The `errors` element will always be an array and contain one or more strings of why the user's request was unsuccessful. Examples will include a "ID was invalid" in the case of a 404, or "the 'description' parameter was missing"
+
+
+## RESTful Endpoints, Minimum Requirements:
 
-You will need to expose Restful API endpoints for both **Items** and **Merchants**.
+You will need to expose the following RESTful API endpoints for the following:
 
-The endpoints you need to expose are:
+* Merchants:
+  * get all merchants, a maximum of 20 at a time
+  * get one merchant
+  * get all items for a given merchant ID
+* Items:
+  * get all items, a maximum of 20 at a time
+  * get one item
+  * create an item
+  * edit an item
+  * delete an item
+  * get the merchant data for a given item ID
 
-### Index of Resource
+## Non-RESTful Endpoints, Minimum Requirements:
 
-This endpoint renders a JSON representation of all records of the requested resource.
+You will get to choose from the following list:
 
-The URI should follow this pattern: `GET /api/v1/<resource>`
+* ONE of following endpoint pairs:
+  * find one MERCHANT based on search criteria AND find all ITEMS based on search criteria
+  * OR:
+  * find one ITEM based on search criteria AND find all MERCHANTS based on search criteria
+
+* FOUR of the following endpoints:
+  * find a quantity of merchants sorted by descending revenue
+  * find a quantity of merchants sorted by descending items sold
+  * total revenue generated in the whole system over a start/end date range
+  * total revenue for a given merchant
+  * find a quantity of items sorted by descending revenue
+  * total revenue of successful invoices which have not yet been shipped
+  * revenue report, broken down by month in ascending date order
+
+In total, the MINIMUM requirement will be 15 endpoints.
+
+**HINT**: Invoices must have a successful transaction and shipped to the customer to be considered as revenue.
```

Comment:
Also as part of a hint do we want to tell them for the start/end date range they should use the dates on the invoice table?

Reply:
no, I think that's something they can still discover on their own